### PR TITLE
[Gluten-522] Add spotless for automatically fixing style issues.

### PR DESCRIPTION
**Pull request by baibaichen**
## What changes were proposed in this pull request?
This is for  https://github.com/oap-project/gluten/issues/522

###  Add spotless-plugin for scala

Update scala-fmt(copied from flink), and add spotless-plugin, so we can fix style issues by `mvn spotless:apply`:
-    Still use scala-style, because there are some corner cases that scala-fmt can't detect, currently, it is  IfBraceChecker.
-    scala-fmt can't detect unused imports.
-    I apply the format for backends-clickhouse, spark-sql-columnar-shims-common, spark-sql-columnar-shims-spark32, and **spark-sql-columnar-shims-spark33**
-    Disabling spotless-java, will be opened in the followup PR by other person

### Simplify POM
1. Remove incremental-scala-compiler, since it's active by default
2. Unified POM name prefix to Gluten and remove duplicate dependencies
3. Fix an issue that <scalatest-maven-plugin.version>  is undefined.
4. Bump <scala.version> to 2.12.15 for spark 3.2.2 to fix the compiler warning.
5. Bump junit to 4.13.1 and define its version in parent POM
6. Refactor \<scala-maven-plugin\>
7. Move \<protobuf.version\> definition in parent POM, and introduce <protobuf.substrait.version> which will be used in substrait-spark
8. 

## How was this patch tested?
using existed Unit test

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -36,10 +36,12 @@
 
   <properties>
     <spark32.version>3.2.2</spark32.version>
+    <spark32.scala>2.12.15</spark32.scala>
     <spark33.version>3.3.0</spark33.version>
     <delta2.version>2.0.0</delta2.version>
     <delta1.version>1.2.1</delta1.version>
     <scala.version>2.12.15</scala.version>
+    <scala.version>${spark32.scala}</scala.version>
```
Comment (zzcclp) [inline pom.xml:44]: duplicate `scala.version`

File: pom.xml [hunk 2]
```diff
@@ -36,10 +36,12 @@
 
   <properties>
     <spark32.version>3.2.2</spark32.version>
+    <spark32.scala>2.12.15</spark32.scala>
```
Comment (zzcclp) [inline pom.xml:39]: maybe only using `scala.version` is enough, and change the version for each spark version in the profiles.

File: pom.xml [hunk 3]
```diff
@@ -488,10 +564,25 @@
...
           </configuration>
+          <executions>
+            <execution>
+              <phase>validate</phase>
+              <goals>
+                <goal>check</goal>
+              </goals>
+            </execution>
+          </executions>
         </plugin>
```
Comment (zhztheplayer) [inline pom.xml:586]: Any reason we will still keep the usage of scalastyle while scalafmt is being introduced?

**Comment by zzcclp** [comment]
Please create a issue and modify the title to `[XXXX-XXX] XXXX`

File: pom.xml (hunk 2 above)
Comment (baibaichen) [reply pom.xml:39]: spark master already bumps to 2.12.17(https://github.com/apache/spark/pull/37892),

Comment (zzcclp) [reply pom.xml:39]: Spark 3.3 still uses 2.12.15, and also can set the different version in the each spark version profile

Comment (baibaichen) [reply pom.xml:39]: I add a property <spark33.scala>

File: shims/spark33/pom.xml [hunk 4]
```diff
@@ -50,13 +50,18 @@
...
-        <artifactId>scalastyle-maven-plugin</artifactId>
-      </plugin>
+      <!--      <plugin>-->
+      <!--        <groupId>org.scalastyle</groupId>-->
+      <!--        <artifactId>scalastyle-maven-plugin</artifactId>-->
+      <!--        <version>${scalastyle.version}</version>-->
+      <!--        <configuration>-->
+      <!--          <configLocation>${project.basedir}/../../scalastyle-config.xml</configLocation>-->
+      <!--        </configuration>-->
+      <!--      </plugin>-->
```
[the remark this answers is not in the document]
Comment (baibaichen) [reply shims/spark33/pom.xml]: The Scala style checker didn't work before this patch, however, there are some violations, for example, the license header doesn't match the apache license.

With the new commits I fix these issues and uncomment them.

File: pom.xml (hunk 1 above)
Comment (baibaichen) [reply pom.xml]: thanks for pointing this

**Comment by rui-mo** [comment]
We found below error in Jenkins test.

[ERROR] Failed to execute goal com.diffplug.spotless:spotless-maven-plugin:2.27.2:check (spotless-check) on project spark-sql-columnar-shims-common: Execution spotless-check of goal com.diffplug.spotless:spotless-maven-plugin:2.27.2:check failed: Unable to resolve dependencies: org.scalameta:parsers_2.12:jar:4.5.12 was not found in http://maven.aliyun.com/nexus/content/groups/public/ during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of alimaven has elapsed or updates are forced -> [Help 1]

**Comment by baibaichen** [comment]
Who can help me on this, it looks like an enviorment issue.

> We found below error in Jenkins test during compilation.
> 
> [ERROR] Failed to execute goal com.diffplug.spotless:spotless-maven-plugin:2.27.2:check (spotless-check) on project spark-sql-columnar-shims-common: Execution spotless-check of goal com.diffplug.spotless:spotless-maven-plugin:2.27.2:check failed: Unable to resolve dependencies: org.scalameta:parsers_2.12:jar:4.5.12 was not found in http://maven.aliyun.com/nexus/content/groups/public/ during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of alimaven has elapsed or updates are forced -> [Help 1]

**Comment by rui-mo** [comment]
> Who can help me on this, it looks like an enviorment issue.
> 
> > We found below error in Jenkins test during compilation.
> > [ERROR] Failed to execute goal com.diffplug.spotless:spotless-maven-plugin:2.27.2:check (spotless-check) on project spark-sql-columnar-shims-common: Execution spotless-check of goal com.diffplug.spotless:spotless-maven-plugin:2.27.2:check failed: Unable to resolve dependencies: org.scalameta:parsers_2.12:jar:4.5.12 was not found in http://maven.aliyun.com/nexus/content/groups/public/ during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of alimaven has elapsed or updates are forced -> [Help 1]

Will try to fix this issue on Jenkins after the rebase. Thanks.

File: pom.xml [hunk 5]
```diff
@@ -494,24 +571,52 @@
...
+<!--                <style>AOSP</style>-->
+<!--              </googleJavaFormat>-->
+
+<!--              &lt;!&ndash; \# refers to the static imports &ndash;&gt;-->
+<!--              <importOrder>-->
+<!--                <order>io.glutenproject,io.substrait.spark,,javax,java,scala,\#</order>-->
+<!--              </importOrder>-->
+
+<!--              <removeUnusedImports />-->
+<!--            </java>-->
```
Comment (zhztheplayer) [inline pom.xml:591]: Any reason commenting this out?

File: pom.xml [hunk 6]
```diff
@@ -383,31 +427,26 @@
...
         <groupId>com.fasterxml.jackson.datatype</groupId>
         <artifactId>jackson-datatype-guava</artifactId>
         <version>${fasterxml.version}</version>
-        <scope>test</scope>
       </dependency>
       <dependency>
         <groupId>com.fasterxml.jackson.module</groupId>
         <artifactId>jackson-module-scala_${scala.binary.version}</artifactId>
         <version>${fasterxml.version}</version>
-        <scope>test</scope>
```
Comment (zhztheplayer) [inline pom.xml:410]: Is this going to bringing jackson to the bundle?

File: .scalafmt.conf [hunk 7]
```diff
@@ -1,11 +1,71 @@
...
+newlines.afterCurlyLambdaParams = squash # No newline after lambda params
+newlines.inInterpolation = "avoid"
+newlines.avoidInResultType = true
+optIn.annotationNewlines = true
+
+# Scaladoc
+docstrings.style = Asterisk # Javadoc style
+docstrings.removeEmpty = true
+docstrings.oneline = fold
+docstrings.forceBlankLineBefore = true
```
[the remark this answers is not in the document]
Comment (baibaichen) [reply .scalafmt.conf:39]: from flink

File: pom.xml (hunk 6 above)
Comment (baibaichen) [reply pom.xml:410]: substrait-core need jackson

File: backends-clickhouse/pom.xml [hunk 8]
```diff
@@ -270,26 +239,22 @@
...
-                </executions>
+            </plugin>
+            <plugin>
+                <groupId>com.diffplug.spotless</groupId>
+                <artifactId>spotless-maven-plugin</artifactId>
+                <version>${spotless.version}</version>
+                <configuration>
+                    <scala>
+                        <scalafmt>
+                            <file>${project.basedir}/../.scalafmt.conf</file>
```
Comment (zhztheplayer) [inline backends-clickhouse/pom.xml:254]: Is it possible to put all format configs `.scalafmt.conf` `.clang-format` `scalastyle-config.xml` into somewhere for example the `dev/` folder? It's not a blocker to the PR and we may raise another patch for that.

File: .scalafmt.conf [hunk 9]
```diff
@@ -1,11 +1,71 @@
...
+rewrite.imports.sort = scalastyle
+rewrite.imports.groups = [
+    ["io.glutenproject\\..*"],
+    ["io.substrait.spark\\..*"],
+    ["org.apache.spark\\..*"],
+    [".*"],
+    ["javax\\..*"],
+    ["java\\..*"],
+    ["scala\\..*"]
+]
```
[the remark this answers is not in the document]
Comment (baibaichen) [reply .scalafmt.conf:57]: intellij also supprots scala fmts, probably using scala  fmt is enough？

File: backends-clickhouse/pom.xml (hunk 8 above)
Comment (baibaichen) [reply backends-clickhouse/pom.xml:254]: sure, let's create an issue for it

Comment (baibaichen) [reply backends-clickhouse/pom.xml:254]: Create https://github.com/oap-project/gluten/issues/530

File: .scalafmt.conf (hunk 9 above)
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply .scalafmt.conf:57]: > intellij also supprots scala fmts, probably using scala fmt is enough？

Yes I tried and we can update `dev/intellij-codestyle.xml` to
```
...
  <ScalaCodeStyleSettings>
    <option name="SCALAFMT_CONFIG_PATH" value="./.scalafmt.conf" />
    <option name="FORMATTER" value="1" />
  </ScalaCodeStyleSettings>
...
```

**Review by zhztheplayer — APPROVED** [review]
Thanks for taking this up @baibaichen !

File: .scalafmt.conf (hunk 9 above)
[the remark this answers is not in the document]
Comment (baibaichen) [reply .scalafmt.conf:57]: Let's do this  in the follow up PR which I'm going to apply spotless on weekend.

**Comment by zhouyuan** [comment]
it seems this patch revert a recent change on h2 database, will need to add back otherwise there is a security issue

```
            <groupId>com.h2database</groupId>
            <artifactId>h2</artifactId>
            <version>1.4.195</version>
```

**Pull request merged**